File: Algorithms/p105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/p105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        index = {}
        for i in range(len(inorder)):
            index[inorder[i]] = i
        return self.buildTreeHelper(preorder,0,len(preorder),inorder,0,len(inorder),index)

    def buildTreeHelper(self, preorder, l1, r1, inorder, l2, r2, index):
        if l1>=r1:
            return None
        i = index[preorder[l1]]
        root = TreeNode(inorder[i])
        root.left = self.buildTreeHelper(preorder,l1+1,l1+1+i-l2,inorder,l2,i,index)
        root.right = self.buildTreeHelper(preorder,l1+1+i-l2,r1,inorder,i+1,r2,index)
        return root

File: Algorithms/p105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/p105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int,int> index;
        for(int i=0;i<inorder.size();++i) index[inorder[i]] = i;
        return buildTreeHelper(preorder,0,preorder.size(),inorder,0,inorder.size(),index);
    }
    TreeNode* buildTreeHelper(vector<int>& preorder, int l1, int r1, vector<int>& inorder, int l2, int r2, unordered_map<int,int>& index) {
        if(l1>=r1) return 0;
        int i = index[preorder[l1]];
        TreeNode* root = new TreeNode(inorder[i]);
        root->left = buildTreeHelper(preorder,l1+1,l1+1+i-l2,inorder,l2,i,index);
        root->right = buildTreeHelper(preorder,l1+1+i-l2,r1,inorder,i+1,r2,index);
        return root;
    }
};
